File: 16/solution.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"log"
	"os"
	"regexp"
	"strconv"
	"strings"
)

type field struct {
	minA, maxA int
	minB, maxB int
	name       string
}

func parseFields(lines []string) []field {
	var result []field
	expression := regexp.MustCompile(`^((?:\w|\s)+)\:\s(\d+)\-(\d+)\sor\s(\d+)\-(\d+)`)
	for _, line := range lines {
		match := expression.FindAllStringSubmatch(line, -1)[0]
		var ints []int
		for _, str := range match[2:] {
			val, err := strconv.Atoi(str)
			if err != nil {
				log.Fatal(err)
			}
			ints = append(ints, val)
		}
		result = append(result, field{
			name: match[1],
			minA: ints[0],
			maxA: ints[1],
			minB: ints[2],
			maxB: ints[3]})
	}
	return result
}

func intSliceIntersect(sliceA []int, sliceB []int) []int {
	var result []int
	for _, valA := range sliceA {
		for _, valB := range sliceB {
			if valA == valB {
				result = append(result, valA)
			}
		}
	}
	return result
}

func valueFitsField(value int, f field) bool {
	return ((value >= f.minA && value <= f.maxA) || (value >= f.minB && value <= f.maxB))
}

func valueFitsFieldSlice(value int, fieldSlice []field) bool {
	for _, f := range fieldSlice {
		if valueFitsField(value, f) {
			return true
		}
	}
	return false
}

// https://stackoverflow.com/a/19374861
func difference(slice1 []int, slice2 []int) []int {
	var diff []int

	// Loop two times, first to find slice1 strings not in slice2,
	// second loop to find slice2 strings not in slice1
	for i := 0; i < 2; i++ {
		for _, s1 := range slice1 {
			found := false
			for _, s2 := range slice2 {
				if s1 == s2 {
					found = true
					break
				}
			}
			// String not found. We add it to return slice
			if !found {
				diff = append(diff, s1)
			}
		}
		// Swap the slices, only if it was the first loop
		if i == 0 {
			slice1, slice2 = slice2, slice1
		}
	}

	return diff
}

func part1(lines []string) int {
	var fieldEndIndex, nearbyTicketsStartIndex int
	for index, line := range lines {
		if line == "your ticket:" {
			fieldEndIndex = index - 1
			nearbyTicketsStartIndex = index + 4
			break
		}
	}
	fields := parseFields(lines[:fieldEndIndex])
	sum := 0
	for _, line := range lines[nearbyTicketsStartIndex:] {
		values := strings.Split(line, ",")
		for _, value := range values {
			valueInt, err := strconv.Atoi(value)
			if err != nil {
				log.Fatal(err)
			}
			if !valueFitsFieldSlice(valueInt, fields) {
				sum += valueInt
			}
		}
	}

	return sum
}

func part2(lines []string) int {
	var fieldEndIndex, yourTicketIndex, nearbyTicketsStartIndex int
	for index, line := range lines {
		if line == "your ticket:" {
			fieldEndIndex = index - 1
			yourTicketIndex = index + 1
			nearbyTicketsStartIndex = index + 4
			break
		}
	}

	fields := parseFields(lines[:fieldEndIndex])
	// find all valid values
	var cleanNearbyValues [][]int
	for _, line := range lines[nearbyTicketsStartIndex:] {
		values := strings.Split(line, ",")
		var cleanValueSlice []int
		for _, value := range values {
			valueInt, err := strconv.Atoi(value)
			if err != nil {
				log.Fatal(err)
			}
			if valueFitsFieldSlice(valueInt, fields) {
				cleanValueSlice = append(cleanValueSlice, valueInt)
			}
		}
		cleanNearbyValues = append(cleanNearbyValues, cleanValueSlice)
	}

	// init possible fields for columns
	var columnPossibleFields [][]int
	for i := 0; i < len(cleanNearbyValues[0]); i++ {
		temp := make([]int, len(fields))
		for j := 0; j < len(fields); j++ {
			temp[j] = j
		}
		columnPossibleFields = append(columnPossibleFields, temp)
	}
	// check possible fields for columns
	for _, ticket := range cleanNearbyValues {
		for columnIndex, column := range ticket {
			var fittingFields []int
			for fieldIndex, f := range fields {
				if valueFitsField(column, f) {
					fittingFields = append(fittingFields, fieldIndex)
				}
			}
			columnPossibleFields[columnIndex] = intSliceIntersect(columnPossibleFields[columnIndex], fittingFields)
		}
	}

	// match columns to fields
	result := make(map[int]string)
	for _, columnA := range columnPossibleFields {
		for index, columnB := range columnPossibleFields {
			if len(columnB) == 1 {
				result[index] = fields[columnB[0]].name
			} else if len(columnB) == len(columnA)+1 {
				result[index] = fields[difference(columnA, columnB)[0]].name
			}
		}
	}

	// calculate final answer
	sum := 1
	yourTicketSplit := strings.Split(lines[yourTicketIndex], ",")
	for columnIndex, name := range result {
		if len(name) >= 9 && name[:9] == "departure" {
			val, err := strconv.Atoi(yourTicketSplit[columnIndex])
			if err != nil {
				log.Fatal(err)
			}
			sum *= val
		}
	}
	return sum
}

func main() {
	flag.Parse()
	if len(flag.Args()) == 0 {
		log.Fatal("Provide the path to input")
	}
	file, err := os.Open(flag.Args()[0])
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	var lines []string
	for scanner.Scan() {
		line := scanner.Text()
		lines = append(lines, line)
	}

	fmt.Printf("Part 1 answer:%v\nPart 2 answer:%v\n", part1(lines), part2(lines))
}

File: 16/solution.py
import sys
import re
import typing


class Field:
    def __init__(self,
                 range_a: typing.Tuple[int, int],
                 range_b: typing.Tuple[int, int],
                 name: str):
        self.min_a = int(range_a[0])
        self.max_a = int(range_a[1])
        self.min_b = int(range_b[0])
        self.max_b = int(range_b[1])
        self.name = name


def parse_fields(lines):
    result = []
    expression = re.compile(r"^((?:\w|\s)+)\:\s(\d+)\-(\d+)\sor\s(\d+)\-(\d+)")
    for line in lines:
        match = expression.findall(line)
        result.append(Field((match[0][1], match[0][2]),
                            (match[0][3], match[0][4]), match[0][0]))
    return result


def value_fits_field(value, field):
    return ((field.min_a <= value <= field.max_a) or
            (field.min_b <= value <= field.max_b))


def value_fits_field_list(value, field_list):
    for field in field_list:
        if value_fits_field(value, field):
            return True
    return False


def part1(lines):
    field_end_index = 0
    nearby_tickets_start = 0
    for index, line in enumerate(lines):
        if line == "your ticket:\n":
            field_end_index = index - 1
            nearby_tickets_start = index + 4
            break

    fields = parse_fields(lines[:field_end_index])
    result_sum = 0
    for line in lines[nearby_tickets_start:]:
        values = line.split(",")
        for val in values:
            if not value_fits_field_list(int(val), fields):
                result_sum += int(val)
    return result_sum


def part2(lines):
    field_end_index = 0
    nearby_tickets_start = 0
    your_ticket_index = 0
    for index, line in enumerate(lines):
        if line == "your ticket:\n":
            field_end_index = index - 1
            your_ticket_index = index + 1
            nearby_tickets_start = index + 4
            break

    fields = parse_fields(lines[:field_end_index])
    clean_tickets = []
    for line in lines[nearby_tickets_start:]:
        values = line.split(",")
        clean_ticket = []
        for val in values:
            if value_fits_field_list(int(val), fields):
                clean_ticket.append(int(val))
        clean_tickets.append(clean_ticket)

    column_possible_fields = [range(len(fields))
                              for j in range(len(clean_tickets[0]))]

    for ticket in clean_tickets:
        for comlumn_index, column in enumerate(ticket):
            fitting_fields = []
            for field_index, field in enumerate(fields):
                if value_fits_field(column, field):
                    fitting_fields.append(field_index)
            column_possible_fields[comlumn_index] = list(
                set(column_possible_fields[comlumn_index]) & set(fitting_fields))

    result = {}
    for column_a in column_possible_fields:
        for index, column_b in enumerate(column_possible_fields):
            if len(column_b) == 1:
                result[index] = fields[column_b[0]].name
            elif len(column_b) == len(column_a) + 1:
                result[index] = fields[list(
                    list(set(column_a)-set(column_b)) + list(set(column_b)-set(column_a)))[0]].name

    result_sum = 1
    your_ticket_split = lines[your_ticket_index].split(",")
    for index, name in result.items():
        if len(name) >= 9 and name[:9] == "departure":
            result_sum *= int(your_ticket_split[index])
    return result_sum


try:
    f = open(sys.argv[1], "r")
except IOError:
    print("Error opening the file, try again")
    sys.exit(1)
with f:
    rawLines = f.readlines()
    f.close()
    # digitArray = [int(s) for s in lines]
    print(
        f"Part 1 answer:{part1(rawLines)} Part 2 answer: {part2(rawLines)}")
